### PR TITLE
feat: add pattern annotations

### LINKED ISSUE #752: Add `@Pattern` annotations to methods that accept arguments representing tag names
Should be using a nice meta annotation we can document neatly.

### PR DESCRIPTION
**Pull request by powercasgamer**
closes #752 

image

### CODE REVIEW

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagInternals.java [hunk 1]
```diff
@@ -36,7 +36,8 @@
  */
 @ApiStatus.Internal
 public final class TagInternals {
-  private static final Pattern TAG_NAME_PATTERN = Pattern.compile("[!?#]?[a-z0-9_-]*");
+  public static final String TAG_NAME = "[!?#]?[a-z0-9_-]*";
```
Comment (kezz) [inline text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagInternals.java]: Should be given a better name, perhaps `TAG_NAME_REGEX`? Also should probably be documented.

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagInternals.java [hunk 2]
```diff
@@ -48,7 +49,7 @@ private TagInternals() {
    * @param tagName the name of the tag
    * @since 4.10.0
    */
-  public static void assertValidTagName(final @NotNull String tagName) {
+  public static void assertValidTagName(@org.intellij.lang.annotations.Pattern(TAG_NAME) final @NotNull String tagName) {
```
Comment (kezz) [inline text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagInternals.java]: Can we avoid this ugly import my making a meta annotation for this? I'm pretty sure the `Pattern` annotation supports being used on another annotation.

**Review by kezz — CHANGES_REQUESTED** [review]

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagInternals.java (hunk 2 above)
Comment (powercasgamer) [reply text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagInternals.java]: Added a annotation in [`64e9571` (#920)](https://github.com/KyoriPowered/adventure/pull/920/commits/64e9571df11c29349274c6268ee953ba829f1340) although I'm not sure of the name and if it's in the right package

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagPattern.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.LOCAL_VARIABLE;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+
+@Documented
+@Pattern(TagInternals.TAG_NAME_REGEX)
+@Retention(RetentionPolicy.CLASS)
+@Target({ METHOD, FIELD, PARAMETER, LOCAL_VARIABLE }) //Unsure of what is needed here
```
Comment (powercasgamer) [inline text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/internal/TagPattern.java]: I'm not sure which of these, if any are needed or if I should just copy the ones the `@Pattern` annotation uses

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/TagPattern.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * A {@link Tag} name must match this pattern. This is used to validate tag names. Uses <b>[!?#]?[a-z0-9_-]*</b>
+ *
+ * @since 4.14.0
+ */
+@Documented
+@Retention(RetentionPolicy.CLASS)
+@Target({ METHOD, FIELD, PARAMETER, LOCAL_VARIABLE })
+public @Pattern(TagInternals.TAG_NAME_REGEX) @interface TagPattern {
```
Comment (kashike) [inline text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/TagPattern.java:46]:
```suggestion
@Documented
@Pattern(TagInternals.TAG_NAME_REGEX) 
@Retention(RetentionPolicy.CLASS)
@Target({ METHOD, FIELD, PARAMETER, LOCAL_VARIABLE })
public @interface TagPattern {
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/TagPattern.java (hunk 4 above)
Comment (powercasgamer) [reply text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/TagPattern.java:46]: If I do this it won't compile due to checkstyle. [discord message](https://discord.com/channels/319355592605040642/923037274176913409/1105769998733029418)

**Review by kashike — APPROVED** [review]

**Pull request merged**